[TORCH_LIBRARY] Add impl_abstract_pystub

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #109529




We want users to be able to define custom ops in C++ but put the
abstract impl in Python (since it is easier to write them in Python and
the abstract impl better models device semantics and data-dependent
operators).

`m.impl_abstract_pystub(opname, python_module, context)` declares the
abstract_impl of the operator to exist in the given python module.
When the abstract_impl needs to be accessed (either via FakeTensor or
Meta), and it does not exist, the PyTorch Dispatcher will yell
with a descriptive error message.

Some details:
- We construct a new global AbstractImplPyStub mapping in
  Dispatcher.cpp. Read/write to this map is protected by the Dispatcher
  lock.
- We add a new Meta Tensor fallback kernel. The fallback errors out if there is
  no meta kernel, but also offers a nicer error message if we see that there is
  a pystub.
- We create a `torch._utils_internal.throw_abstract_impl_not_imported_error`
  helper function to throw errors. This way, we can throw different error
  messages in OSS PyTorch vs internal PyTorch. To invoke this from C++, we
  added a PyInterpreter::throw_abstract_impl_not_imported_error.

Differential Revision: [D49464753](https://our.internmc.facebook.com/intern/diff/D49464753/)

Differential Revision: [D49464753](https://our.internmc.facebook.com/intern/diff/D49464753)

zou3519 committed:
[TORCH_LIBRARY] Add impl_abstract_pystub

We want users to be able to define custom ops in C++ but put the
abstract impl in Python (since it is easier to write them in Python and
the abstract impl better models device semantics and data-dependent
operators).

`m.impl_abstract_pystub(opname, python_module)` declares the
abstract_impl of the operator to exist in the given python module.
When the abstract_impl needs to be accessed (either via FakeTensor or
Meta), and it does not exist, the PyTorch Dispatcher will lazily import
the specified `python_module`.

Some details:
- We construct a new global AbstractImplPyStub mapping in
  Dispatcher.cpp. Read/write to this map is protected by the Dispatcher
  lock.
- We add a new Meta Tensor fallback kernel. The fallback (1) attempts to
  import any specified pystubs and (2) errors out if the abstract impl
  doesn't exist.
- Before this PR, FakeTensor looked up the abstract impl from a global
  dict. Now, it will attempt to also lazily import the module specified
  by the pystub if it exists.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "[TORCH_LIBRARY] Add impl_abstract_pystub"


We want users to be able to define custom ops in C++ but put the
abstract impl in Python (since it is easier to write them in Python and
the abstract impl better models device semantics and data-dependent
operators).

`m.impl_abstract_pystub(opname, python_module)` declares the
abstract_impl of the operator to exist in the given python module.
When the abstract_impl needs to be accessed (either via FakeTensor or
Meta), and it does not exist, the PyTorch Dispatcher will lazily import
the specified `python_module`.

Some details:
- We construct a new global AbstractImplPyStub mapping in
  Dispatcher.cpp. Read/write to this map is protected by the Dispatcher
  lock.
- We add a new Meta Tensor fallback kernel. The fallback (1) attempts to
  import any specified pystubs and (2) errors out if the abstract impl
  doesn't exist.
- Before this PR, FakeTensor looked up the abstract impl from a global
  dict. Now, it will attempt to also lazily import the module specified
  by the pystub if it exists.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "[TORCH_LIBRARY] Add impl_abstract_pystub"


We want users to be able to define custom ops in C++ but put the
abstract impl in Python (since it is easier to write them in Python and
the abstract impl better models device semantics and data-dependent
operators).

`m.impl_abstract_pystub(opname, python_module)` declares the
abstract_impl of the operator to exist in the given python module.
When the abstract_impl needs to be accessed (either via FakeTensor or
Meta), and it does not exist, the PyTorch Dispatcher will lazily import
the specified `python_module`.

Some details:
- We construct a new global AbstractImplPyStub mapping in
  Dispatcher.cpp. Read/write to this map is protected by the Dispatcher
  lock.
- We add a new Meta Tensor fallback kernel. The fallback (1) attempts to
  import any specified pystubs and (2) errors out if the abstract impl
  doesn't exist.
- PyInterpreter gets a new "op_registration_import" API that is callable
  from C++.
- Before this PR, FakeTensor looked up the abstract impl from a global
  dict. Now, it will attempt to also lazily import the module specified
  by the pystub if it exists.

Test Plan:
- new tests. But tests are kind of difficult to write since we need to
  dynamically load a C++ extension.

[ghstack-poisoned]

zou3519 committed:
Update on "[TORCH_LIBRARY] Add impl_abstract_pystub"


We want users to be able to define custom ops in C++ but put the
abstract impl in Python (since it is easier to write them in Python and
the abstract impl better models device semantics and data-dependent
operators).

`m.impl_abstract_pystub(opname, python_module)` declares the
abstract_impl of the operator to exist in the given python module.
When the abstract_impl needs to be accessed (either via FakeTensor or
Meta), and it does not exist, the PyTorch Dispatcher will lazily import
the specified `python_module`.

Some details:
- We construct a new global AbstractImplPyStub mapping in
  Dispatcher.cpp. Read/write to this map is protected by the Dispatcher
  lock.
- We add a new Meta Tensor fallback kernel. The fallback (1) attempts to
  import any specified pystubs and (2) errors out if the abstract impl
  doesn't exist.
- PyInterpreter gets a new "op_registration_import" API that is callable
  from C++.
- Before this PR, FakeTensor looked up the abstract impl from a global
  dict. Now, it will attempt to also lazily import the module specified
  by the pystub if it exists.

Test Plan:
- new tests. But tests are kind of difficult to write since we need to
  dynamically load a C++ extension.

[ghstack-poisoned]

zou3519 committed:
Update on "[TORCH_LIBRARY] Add impl_abstract_pystub"


We want users to be able to define custom ops in C++ but put the
abstract impl in Python (since it is easier to write them in Python and
the abstract impl better models device semantics and data-dependent
operators).

`m.impl_abstract_pystub(opname, python_module)` declares the
abstract_impl of the operator to exist in the given python module.
When the abstract_impl needs to be accessed (either via FakeTensor or
Meta), and it does not exist, the PyTorch Dispatcher will lazily import
the specified `python_module`.

Some details:
- We construct a new global AbstractImplPyStub mapping in
  Dispatcher.cpp. Read/write to this map is protected by the Dispatcher
  lock.
- We add a new Meta Tensor fallback kernel. The fallback (1) attempts to
  import any specified pystubs and (2) errors out if the abstract impl
  doesn't exist.
- PyInterpreter gets a new "op_registration_import" API that is callable
  from C++.
- Before this PR, FakeTensor looked up the abstract impl from a global
  dict. Now, it will attempt to also lazily import the module specified
  by the pystub if it exists.

Test Plan:
- new tests. But tests are kind of difficult to write since we need to
  dynamically load a C++ extension.

[ghstack-poisoned]

zou3519 committed:
Update on "[TORCH_LIBRARY] Add impl_abstract_pystub"


We want users to be able to define custom ops in C++ but put the
abstract impl in Python (since it is easier to write them in Python and
the abstract impl better models device semantics and data-dependent
operators).

`m.impl_abstract_pystub(opname, python_module)` declares the
abstract_impl of the operator to exist in the given python module.
When the abstract_impl needs to be accessed (either via FakeTensor or
Meta), and it does not exist, the PyTorch Dispatcher will lazily import
the specified `python_module`.

Some details:
- We construct a new global AbstractImplPyStub mapping in
  Dispatcher.cpp. Read/write to this map is protected by the Dispatcher
  lock.
- We add a new Meta Tensor fallback kernel. The fallback (1) attempts to
  import any specified pystubs and (2) errors out if the abstract impl
  doesn't exist.
- PyInterpreter gets a new "op_registration_import" API that is callable
  from C++.
- Before this PR, FakeTensor looked up the abstract impl from a global
  dict. Now, it will attempt to also lazily import the module specified
  by the pystub if it exists.

Test Plan:
- new tests. But tests are kind of difficult to write since we need to
  dynamically load a C++ extension.

[ghstack-poisoned]

ezyang commented on torch/utils/_python_dispatch.py:
```diff
@@ -380,3 +381,14 @@ def get_arg_idx_from_alias(output_alias):
+        raise ModuleNotFoundError(
+            f'Could not import "{module}". The operator "{context}" '
+            f"specified that its abstract impl is located in that module."
+            f"Please check that the module is available and/or installed."
```
lint prolly gonna complain here

ezyang commented on aten/src/ATen/core/dispatch/Dispatcher.cpp:
```diff
@@ -254,6 +255,66 @@ void Dispatcher::deregisterDef_(
+namespace {
+
+using AbstractImplPyStubsType = std::unordered_map<at::OperatorName, const char*>;
+AbstractImplPyStubsType& kAbstractImplPyStubs() {
```
Why k prefix 🤔

ezyang commented on aten/src/ATen/core/dispatch/Dispatcher.cpp:
```diff
@@ -254,6 +255,66 @@ void Dispatcher::deregisterDef_(
+      "the abstract impl for this operator (necessary for Meta Tensors) "
+      "was declared to exist in the Python module ", pymodule,
+      " but Python is not available.");
+  (*interpreter)->op_registration_pyimport(pymodule, toString(op_name));
```
So in the event of a race, this can trigger multiple times right? (Which should be OK because module imports should be idempotent)

ezyang replied:
I had to check, but it looks like the dispatcher is read-write thread safe, which is *phew*

ezyang replied:
Actually, no, this is kind of sus. Only *part* of the dispatcher is read-write threadsafe, and not the important part.

The critical data structure to protect is

```
std::array<KernelFunction, c10::num_runtime_entries> dispatchTable_;
```

which is what actually gets poked when we actually attempt to do an operator call. The race is when you have two threads simultaneously trying to trigger a meta call: one thread will trigger the pyimport load while the other thread is attempting to do lookup.

KernelFunction is a chonky boy

```
  BoxedKernel boxed_kernel_func_;
  void* unboxed_kernel_func_;
  void* sym_unboxed_kernel_func_;
```

which means that even in x86 TSO semantics it cannot be atomically updated. So you could potentially see a partially invalid KernelFunction in concurrent access.

This problem occurs solely because we are relying on the imported module to directly overwrite the dispatcher entry for Meta upon import, which is the logical and BC way to do things. Can we easily make this thread safe? In general we cannot take out a lock on reads for generic kernels.

One potential fix is to prevent `impl_abstract` from actually registering into the dispatch table directly, but instead some side table that is read-write race safe. The meta fallback redispatches into that table when the entry exists.

Or lmk if you want to YOLO, since in practice you're likely to be guarded by a higher level lock already, the torch.compile lock.

zou3519 replied:
A race could happen if multiple threads invoke DispatchKey::Meta's fallback (so `op_registration_pyimport` could get run by multiple threads). op_registration_pyimport needs to acquire the GIL, so the GIL would make it so that only one thread modifies the Dispatcher state at a time. (also yes, imports are idempotent, so the second time an import happens it no-ops).

zou3519 replied:
Oh I see, you're right. Let me think about this for a bit

ezyang replied:
Oh, an even simpler way to do this is to just forbid Meta registrations (or any other registrations) here (only go into the fake tensor side table). Because I don't think we need the *meta* for anything.

ezyang commented on torch/csrc/utils/python_dispatch.cpp:
```diff
@@ -422,6 +422,17 @@ void initDispatchBindings(PyObject* module) {
 
+  m.def(
+      "_dispatch_maybe_import_abstract_impl_pystub", [](const char* qualname) {
+        // The abstract_impl API doesn't work with overload names yet.
```
:0

zou3519 replied:
The `impl_{blah}` APIs are written in a spaghetti fashion because we originally wanted it to be a new python-only custom op API and then migrated them to work with C++ custom ops without actually doing a refactor.

I'm in the process of doing a refactor to get the `impl_{blah}` APIs to really look like "an API that sits directly above the low-level torch.library DispatchKey APIs"

ezyang reviewed (approved):
Preemptively approving but let's take a close look at the thread safety thing

bdhirsh commented on aten/src/ATen/core/MetaFallbackKernel.cpp:
```diff
@@ -0,0 +1,33 @@
+    const c10::OperatorHandle& op,
+    c10::DispatchKeySet dispatch_keys,
+    torch::jit::Stack* stack) {
+  bool resolved = c10::Dispatcher::singleton().maybeImportAbstractImplPyStub(op.operator_name());
```
What's the relationship now between this and the `torch._custom_op.impl.global_registry`? IIUC, it seems like fake tensor will check both, but if you run vanilla meta tensor logic, we will ignore the global_registry and only try this path.

bdhirsh replied:
Oh nvm - that call will just do the python import, and the user's python module will (ideally) call `.abstract_impl`, populating both the global registry and registering a meta kernel.

I wasn't sure at first whether calling a custom op on the meta device would repeatedly hit this fallback, but that's wrong - the import will populate the dispatcher with a kernel registered directly to the meta key, which will now have higher precedence than the boxed fallback, so redispatching will do the right thing.

bdhirsh replied:
What happens if someone uses `impl_abstract_pystub`, but their python module doesn't properly register a meta kernel to the dispatcher? Will they infinite loop in this fallback if you try to call their op with meta tensors?

bdhirsh commented on torch/csrc/utils/python_dispatch.cpp:
```diff
@@ -422,6 +422,17 @@ void initDispatchBindings(PyObject* module) {
 
+  m.def(
+      "_dispatch_maybe_import_abstract_impl_pystub", [](const char* qualname) {
+        // The abstract_impl API doesn't work with overload names yet.
```
what the stance on overload names? Banned from pure-python custom ops, but need-to-support for existing C++ custom ops?

bdhirsh commented on torch/_custom_op/impl.py:
```diff
@@ -1078,6 +1078,29 @@ def _find_custom_op(qualname, also_check_torch_library=False):
+    # This is a bit slow so we do it after the fast-path above
+    did_import = torch._C._dispatch_maybe_import_abstract_impl_pystub(qualname)
+    if did_import:
+        return get_abstract_impl_from_registry(qualname)
```
Kind of similar to the meta tensor fallback comment - maybe we could raise a slightly better error message here (than the default "cannot find an abstract impl"), if we know the custom op has an `impl_abstract_pystub("my_py_module")` somewhere, but running it didn't result in an abstract impl populating the table?

bdhirsh commented on aten/src/ATen/core/dispatch/Dispatcher.cpp:
```diff
@@ -254,6 +255,66 @@ void Dispatcher::deregisterDef_(
 
+namespace {
+
+using AbstractImplPyStubsType = std::unordered_map<at::OperatorName, const char*>;
```
I think most of `c10` uses `ska::flat_hash_map` over `unordered_map`, although perf probably doesn't matter a ton here.

zou3519 commented:
@zou3519 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D49464753).

zou3519 commented on aten/src/ATen/core/MetaFallbackKernel.cpp:
```diff
@@ -0,0 +1,33 @@
+namespace at {
+namespace impl {
+
+static void metaFallback(
```
there aren't per-op fallbacks, right? And there isn't really a way to easily register and de-register kernels

bdhirsh replied:
(technically you can register a boxed function to a single operator) but nvm you're right - the fallback seems like the right thing anyway since you're expecting the python module to eventually register something directly to the operator.

zou3519 committed:
Update on "[TORCH_LIBRARY] Add impl_abstract_pystub"





We want users to be able to define custom ops in C++ but put the
abstract impl in Python (since it is easier to write them in Python and
the abstract impl better models device semantics and data-dependent
operators).

`m.impl_abstract_pystub(opname, python_module, context)` declares the
abstract_impl of the operator to exist in the given python module.
When the abstract_impl needs to be accessed (either via FakeTensor or
Meta), and it does not exist, the PyTorch Dispatcher will yell
with a descriptive error message.

Some details:
- We construct a new global AbstractImplPyStub mapping in
  Dispatcher.cpp. Read/write to this map is protected by the Dispatcher
  lock.
- We add a new Meta Tensor fallback kernel. The fallback errors out if there is
  no meta kernel, but also offers a nicer error message if we see that there is
  a pystub.
- We create a `torch._utils_internal.throw_abstract_impl_not_imported_error`
  helper function to throw errors. This way, we can throw different error
  messages in OSS PyTorch vs internal PyTorch. To invoke this from C++, we
  added a PyInterpreter::throw_abstract_impl_not_imported_error.

Differential Revision: [D49464753](https://our.internmc.facebook.com/intern/diff/D49464753/)

Differential Revision: [D49464753](https://our.internmc.facebook.com/intern/diff/D49464753)

[ghstack-poisoned]

zou3519 committed:
Update on "[TORCH_LIBRARY] Add impl_abstract_pystub"





We want users to be able to define custom ops in C++ but put the
abstract impl in Python (since it is easier to write them in Python and
the abstract impl better models device semantics and data-dependent
operators).

`m.impl_abstract_pystub(opname, python_module, context)` declares the
abstract_impl of the operator to exist in the given python module.
When the abstract_impl needs to be accessed (either via FakeTensor or
Meta), and it does not exist, the PyTorch Dispatcher will yell
with a descriptive error message.

Some details:
- We construct a new global AbstractImplPyStub mapping in
  Dispatcher.cpp. Read/write to this map is protected by the Dispatcher
  lock.
- We add a new Meta Tensor fallback kernel. The fallback errors out if there is
  no meta kernel, but also offers a nicer error message if we see that there is
  a pystub.
- We create a `torch._utils_internal.throw_abstract_impl_not_imported_error`
  helper function to throw errors. This way, we can throw different error
  messages in OSS PyTorch vs internal PyTorch. To invoke this from C++, we
  added a PyInterpreter::throw_abstract_impl_not_imported_error.

Differential Revision: [D49464753](https://our.internmc.facebook.com/intern/diff/D49464753/)

Differential Revision: [D49464753](https://our.internmc.facebook.com/intern/diff/D49464753)

[ghstack-poisoned]

zou3519 commented:
We fixed the locking issues by not actually importing the pystub if it has not been imported yet. Instead, we just raise an error and ask the user to import the pystub.

zou3519 committed:
Update on "[TORCH_LIBRARY] Add impl_abstract_pystub"





We want users to be able to define custom ops in C++ but put the
abstract impl in Python (since it is easier to write them in Python and
the abstract impl better models device semantics and data-dependent
operators).

`m.impl_abstract_pystub(opname, python_module, context)` declares the
abstract_impl of the operator to exist in the given python module.
When the abstract_impl needs to be accessed (either via FakeTensor or
Meta), and it does not exist, the PyTorch Dispatcher will yell
with a descriptive error message.

Some details:
- We construct a new global AbstractImplPyStub mapping in
  Dispatcher.cpp. Read/write to this map is protected by the Dispatcher
  lock.
- We add a new Meta Tensor fallback kernel. The fallback errors out if there is
  no meta kernel, but also offers a nicer error message if we see that there is
  a pystub.
- We create a `torch._utils_internal.throw_abstract_impl_not_imported_error`
  helper function to throw errors. This way, we can throw different error
  messages in OSS PyTorch vs internal PyTorch. To invoke this from C++, we
  added a PyInterpreter::throw_abstract_impl_not_imported_error.

Differential Revision: [D49464753](https://our.internmc.facebook.com/intern/diff/D49464753/)

Differential Revision: [D49464753](https://our.internmc.facebook.com/intern/diff/D49464753)

[ghstack-poisoned]

zou3519 committed:
Update on "[TORCH_LIBRARY] Add impl_abstract_pystub"





We want users to be able to define custom ops in C++ but put the
abstract impl in Python (since it is easier to write them in Python and
the abstract impl better models device semantics and data-dependent
operators).

`m.impl_abstract_pystub(opname, python_module, context)` declares the
abstract_impl of the operator to exist in the given python module.
When the abstract_impl needs to be accessed (either via FakeTensor or
Meta), and it does not exist, the PyTorch Dispatcher will yell
with a descriptive error message.

Some details:
- We construct a new global AbstractImplPyStub mapping in
  Dispatcher.cpp. Read/write to this map is protected by the Dispatcher
  lock.
- We add a new Meta Tensor fallback kernel. The fallback errors out if there is
  no meta kernel, but also offers a nicer error message if we see that there is
  a pystub.
- We create a `torch._utils_internal.throw_abstract_impl_not_imported_error`
  helper function to throw errors. This way, we can throw different error
  messages in OSS PyTorch vs internal PyTorch. To invoke this from C++, we
  added a PyInterpreter::throw_abstract_impl_not_imported_error.

Differential Revision: [D49464753](https://our.internmc.facebook.com/intern/diff/D49464753/)

Differential Revision: [D49464753](https://our.internmc.facebook.com/intern/diff/D49464753)

[ghstack-poisoned]

zou3519 commented:
All the test failures look unrelated, so will force merge

zou3519 committed:
Update on "[TORCH_LIBRARY] Add impl_abstract_pystub"





We want users to be able to define custom ops in C++ but put the
abstract impl in Python (since it is easier to write them in Python and
the abstract impl better models device semantics and data-dependent
operators).

`m.impl_abstract_pystub(opname, python_module, context)` declares the
abstract_impl of the operator to exist in the given python module.
When the abstract_impl needs to be accessed (either via FakeTensor or
Meta), and it does not exist, the PyTorch Dispatcher will yell
with a descriptive error message.

Some details:
- We construct a new global AbstractImplPyStub mapping in
  Dispatcher.cpp. Read/write to this map is protected by the Dispatcher
  lock.
- We add a new Meta Tensor fallback kernel. The fallback errors out if there is
  no meta kernel, but also offers a nicer error message if we see that there is
  a pystub.
- We create a `torch._utils_internal.throw_abstract_impl_not_imported_error`
  helper function to throw errors. This way, we can throw different error
  messages in OSS PyTorch vs internal PyTorch. To invoke this from C++, we
  added a PyInterpreter::throw_abstract_impl_not_imported_error.

Differential Revision: [D49464753](https://our.internmc.facebook.com/intern/diff/D49464753/)

Differential Revision: [D49464753](https://our.internmc.facebook.com/intern/diff/D49464753)

[ghstack-poisoned]

zou3519 committed:
Update on "[TORCH_LIBRARY] Add impl_abstract_pystub"





We want users to be able to define custom ops in C++ but put the
abstract impl in Python (since it is easier to write them in Python and
the abstract impl better models device semantics and data-dependent
operators).

`m.impl_abstract_pystub(opname, python_module, context)` declares the
abstract_impl of the operator to exist in the given python module.
When the abstract_impl needs to be accessed (either via FakeTensor or
Meta), and it does not exist, the PyTorch Dispatcher will yell
with a descriptive error message.

Some details:
- We construct a new global AbstractImplPyStub mapping in
  Dispatcher.cpp. Read/write to this map is protected by the Dispatcher
  lock.
- We add a new Meta Tensor fallback kernel. The fallback errors out if there is
  no meta kernel, but also offers a nicer error message if we see that there is
  a pystub.
- We create a `torch._utils_internal.throw_abstract_impl_not_imported_error`
  helper function to throw errors. This way, we can throw different error
  messages in OSS PyTorch vs internal PyTorch. To invoke this from C++, we
  added a PyInterpreter::throw_abstract_impl_not_imported_error.

Differential Revision: [D49464753](https://our.internmc.facebook.com/intern/diff/D49464753/)

Differential Revision: [D49464753](https://our.internmc.facebook.com/intern/diff/D49464753)

[ghstack-poisoned]
